File: algorithms/08_r/sorting/merge_sort.R
merge_sort <- function(arr) {
  if (length(arr) <= 1) return(arr)
  mid <- length(arr) %/% 2
  left <- merge_sort(arr[1:mid])
  right <- merge_sort(arr[(mid + 1):length(arr)])
  
  res <- c()
  i <- 1
  j <- 1
  while (i <= length(left) && j <= length(right)) {
    if (left[i] <= right[j]) {
      res <- c(res, left[i])
      i <- i + 1
    } else {
      res <- c(res, right[j])
      j <- j + 1
    }
  }
  if (i <= length(left)) res <- c(res, left[i:length(left)])
  if (j <= length(right)) res <- c(res, right[j:length(right)])
  return(res)
}

File: algorithms/12_php/sorting/merge_sort.php
<?php
declare(strict_types=1);

namespace Algorithms\Sorting;

function mergeSort(array $arr): array {
    if (count($arr) <= 1) return $arr;
    $mid = intdiv(count($arr), 2);
    $left = mergeSort(array_slice($arr, 0, $mid));
    $right = mergeSort(array_slice($arr, $mid));
    
    $res = [];
    $i = $j = 0;
    while ($i < count($left) && $j < count($right)) {
        if ($left[$i] <= $right[$j]) {
            $res[] = $left[$i++];
        } else {
            $res[] = $right[$j++];
        }
    }
    return array_merge($res, array_slice($left, $i), array_slice($right, $j));
}
